Fallback assets with the tag "scribble asset"

tabularelf opened the pull request:
In 2024.11, a feature was added to allow removing unused assets in a project. A feature that was enabled by default in new projects. Which caused some issues with Scribble usage.
2024.13 introduced a new `gml_pragma()` that allows marking all assets under a specific tag, to be utilised.

This PR introduces a small change to apply said asset tag on said fallback assets. (Which carries through local packages.)
This allows for the end user to still respect "automatically remove unused assets", while marking specific assets for Scribble use.
By adding `gml_pragma("MarkTagAsUsed", "scribble asset")`, and applying `"scribble asset"` to any of their own assets.

(Note: This PR does include the `gml_pragma()` call. More info down below in a comment.)

tabularelf commented:
As this PR includes `gml_pragma("MarkTagAsUsed", "scribble asset")`, accepting this PR will now prevent the following error, in relation to "Automatically remove unused assets when compiling" being checked in Game Options, as the PR applies changes to the scribble fallback assets.

```
ERROR in action number 1
of Create Event for object <undefined>:
 
Scribble Deluxe 9.3.5:
Fallback font was not found. This may indicate that unused assets have been stripped from the project
Please untick "Automatically remove unused assets when compiling" in Game Options
```

Currently Scribble can only fetch assets reference via string name (`"[spr_npc_bob_icon, 2]"`), however in hopes of this PR being accepted, changes will be applied from discussions held in #cookbook_jam in GM Kitchen. Where Scribble shall support `handle_parse`. Because unlike the current Scribble method of fetching assets, which requires `asset_get_index`, `handle_parse` allows us to directly convert a reference into a string variant, via template strings or similar approaches `$"[{spr_npc_bob_icon}, 2]"` which transform the string into `"[ref sprite spr_npc_bob_icon, 2]"` `handle_parse` can take `ref sprite spr_npc_bob_icon`, as a similar fashion to `asset_get_index`, and return the asset ref handle in real time.

Which means that we can instruct people to instead handle the alternative with scribble as:
```gml
var element = scribble($"[{spr_npc_bob_icon}, 2]Hello {global.player_name}! I hope your adventure was good!");
element.draw(x, y-32);
```
OR
```gml
var element = scribble("[" + string(spr_npc_bob_icon) + ", 2]Hello {global.player_name}! I hope your adventure was good!");
element.draw(x, y-32);
```
OR
```gml
var str = string("[{0}, 2]Hello {global.player_name}! I hope your adventure was good!", spr_npc_bob_icon);
var element = scribble(str);
element.draw(x, y-32);
```

JujuAdams commented:
Cool! Thank you for this. I won't merge the PR just yet, got a few other things to think about first, but this is a positive change overall.

JujuAdams commented:
Not for v9 as it represents a workflow change that'll inevitably catch people out. Maybe for v10 where there's an opportunity to break hearts.

tinkerer-red commented:
i dont see any reason both systems cant exist at once, they would need to be for LTS support at the least, maybe v10 or in the future you could deprecate `[spr_name]`, but implementing this shouldnt force changes for the existing work flow, additionally the same concepts can be applied to textures and any other refs too. But if its being put off until a full implementation then thats a different story, as we dont have surface refs yet.

tabularelf commented:
> i dont see any reason both systems cant exist at once, they would need to be for LTS support at the least, maybe v10 or in the future you could deprecate `[spr_name]`, but implementing this shouldnt force changes for the existing work flow, additionally the same concepts can be applied to textures and any other refs too. But if its being put off until a full implementation then thats a different story, as we dont have surface refs yet.

Problem I mostly see is that people are currently used to how v9 and earlier work. So that making the change right now would at the very least throw people off. Something like this could be tucked away behind a macro config (similar to what chatterbox did when I've introduced `script_execute_ext`), but that would at most be disabled by default.

Beyond that, I think aiming for v10 wouldn't be a bad idea. Gives plenty of time to rework things. 

> note:
> 
> I have not actually looked at Tab's current implementation, i wouldnt imagine he would remove the other tags system but maybe he did.

I changed nothing about the tag system. This PR merely just applies "scribble asset" to key default assets, and allowing Scribble to process them in game start. Which currently breaks the existing checks with detecting for "automatically remove unused assets" being checked.

tinkerer-red commented:
Only really have two small notes related to this.

Asset renaming which while convenient to `$"foo {spr_bar}" which would work with ide refactoring methods, is a moot point when able to find replace all (should be using better naming for sprites if its an issue)

This could co-exist; but of course then you need to increase complexity by checking if the last 11 char were `ref sprite ` and if the next set are actually valid reference pointers, but then this breaks out of the json like bracket structure which leads to even further issues.

at most i could see `$"[{spr_happy}]"` being valid, but even then that should just be optional.

tabularelf commented:
On Tue, 26 Aug 2025 at 8:36 pm, Tinkerer_Red ***@***.***>
wrote:

> *tinkerer-red* left a comment (JujuAdams/Scribble#613)
> <https://github.com/JujuAdams/Scribble/pull/613#issuecomment-3223602235>
>
> Asset renaming which while convenient to `$"foo {spr_bar}" which would
> work with ide refactoring methods, is a moot point when able to find
> replace all (should be using better naming for sprites if its an issue)
>
The issue is really down to external files, that’s the problem. GameMaker
does not check external files at all, so you would need separate tools to
handle those. Anything GML wise isn’t important as the compiler is smart
enough to know (as smart as it is, anyway). Asset renaming wouldn’t
necessarily address the problem. 🤔

This could co-exist; but of course then you need to increase complexity by
> checking if the last 11 char were ref sprite and if the next set are
> actually valid reference pointers, but then this breaks out of the json
> like bracket structure which leads to even further issues.
>
That’s certainly not impossible. Heck just checking for `ref ` at the start
is enough. Fall back to a regular `asset_get_index()` lookup. (But again
that’s a whole separate thing, not what strictly the PR was initially
addressing.)

> at most i could see $"[{spr_happy}]" being valid, but even then that
> should just be optional.
>
Referring to the previous quote, that would be fine?
